From d8b529c65ef435c46c8d3fd73ce946694d57298a Mon Sep 17 00:00:00 2001
From: Arnon Kaufman <akaufman@habana.ai>
Date: Wed, 16 Jun 2021 10:19:53 +0300
Subject: [PATCH] [SW-45908] security: check fit access in range

When checking the incoming FIT validity the size of the fit should fit
into the SRAM valid address space. The same goes with the FIT's footer.
Before accessing the fit or footer the calculated intended address is
checked against the valid SRAM boundaries.

Change-Id: I5d7dbe7a46424e4574a341bfb99581df5b616b66
Signed-off-by: Arnon Kaufman <akaufman@habana.ai>
---

diff --git a/gaudi/security.c b/gaudi/security.c
index bcee20b..bfac2e7 100644
--- a/gaudi/security.c
+++ b/gaudi/security.c
@@ -1868,6 +1868,13 @@
 	/* security descriptor is at the footer of the image */
 	fit_sec_off = (const void *)((u8 *)img + img_size);
 
+	/* check that the calculated memory range are legal */
+	if (!preboot_sram_is_valid_address_range((uintptr_t)fit_sec_off,
+						 sizeof(fit_sec))) {
+		preboot_puts("Invalid address range\n");
+		return 1;
+	}
+
 	memcpy(&fit_sec, fit_sec_off, FIT_SEC_DESC_SIZE);
 
 	if (fit_sec.magic_number != FIT_SEC_MAGIC) {
diff --git a/preboot/fit.c b/preboot/fit.c
index 4cfa225..8c2baf7 100644
--- a/preboot/fit.c
+++ b/preboot/fit.c
@@ -779,6 +779,10 @@
 	if (fdt_version(fit) < FDT_MIN_SUPPORTED_VERSION)
 		return FIT_LOAD_SEC_FAILED;
 
+	/* check that the calculated memory range are legal */
+	if (!preboot_sram_is_valid_address_range((uintptr_t)fit, fit_size))
+		return FIT_LOAD_SEC_FAILED;
+
 	/* Get number of images in preboot configuration */
 	conf = fit_find_config_node(fit);
 	if (conf < 0)
